feat: introduce a Ruin And Recreate move (#1010)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/optimization-algorithms/move-selector-reference.adoc

An advanced configuration looks like this:

[source,xml,options="nowrap"]
----
    <pillar...MoveSelector>
      ...
      <subPillarType>SEQUENCE</subPillarType>
      <subPillarSequenceComparatorClass>...ShiftComparator</subPillarSequenceComparatorClass>
      <pillarSelector>
        ...
        <minimumSubPillarSize>1</minimumSubPillarSize>
        <maximumSubPillarSize>1000</maximumSubPillarSize>
      </pillarSelector>
      ...
    </pillar...MoveSelector>
----

In this case, the entity being operated on needn’t be `Comparable`.
The given `subPillarSequenceComparatorClass` is used to establish the sequence instead.
Also, the size of the sub-pillars is limited in length of up to 1000 entities.

[#ruinRecreateMoveSelector]
=== `RuinRecreateMoveSelector`

The `RuinRecreateMove` selects a subset of entities and sets their values to null,
effectively unassigning them.
Then it runs a construction heuristic to assign them again.
If xref:using-timefold-solver/modeling-planning-problems.adoc#planningVariableAllowingUnassigned[unassigned values] are allowed,
it may leave them unassigned.

> This coarse-grained move is useful to help get the solver out of a local optimum.

Reviewer: ```suggestion
This coarse-grained move is useful to help the solver to escape from a local optimum.
```

----

[#kOptListMoveSelector]
=== `KOptListMoveSelector`

The `KOptListMoveSelector` considers the list variable to be
a graph whose edges are the consecutive elements of the list
(with the last element being consecutive to the first element).
A `KOptListMove` selects an entity, remove `k` edges from its list variable, and add `k` new edges from the removed edges' endpoints.
This move may reverse segments of the graph.

image::optimization-algorithms/move-selector-reference/koptMove.png[align="center"]

Simplest configuration:

[source,xml]
----
    <kOptListMoveSelector/>
----

Advanced configuration:

[source,xml]
----
    <kOptListMoveSelector>
      ... <!-- Normal selector properties -->
      <minimumK>2</minimumK>
      <maximumK>4</maximumK>
    </kOptListMoveSelector>
----

[#listRuinRecreateMoveSelector]
=== `ListRuinRecreateMoveSelector`

The `ListRuinRecreateMove` selects a subset of values, and removes them from their list variables.
Then it runs a construction heuristic to assign them again.
If xref:using-timefold-solver/modeling-planning-problems.adoc#planningListVariableAllowingUnassigned[unassigned values] are allowed,
it may leave them unassigned.

> This coarse-grained move is useful to help get the solver out of a local optimum.

Reviewer: ´´´suggestion
This coarse-grained move is useful to help the solver to escape from a local optimum.
´´´
